### DISCUSSION TITLE
How to upgrade using a bundle that has a new slot?

### DISCUSSION THREAD

**Discussion by erickson-ureta** [rauc/rauc#929] [Q&A]
For my project, I have to add a new slot to our RAUC bundles. The hope was that we could just upgrade our existing systems to simply add the new slot after an upgrade.

RAUC however does not like it when the new bundle's manifest has thew new slot, and complains:
```
Installation error: No target slot for class <newslotname> of image <imagename> found
 ```
Now I know this is solveable via a double upgrade, where the first bundle's rootfs just contains an updated `system.conf` that supports the new slot, then followed by a RAUC bundle that actually contains the new slot info.
The problem with this approach is that we have some units running in our customers' labs, and we would like to avoid telling them to manually fiddle with these units' `system.conf` files or telling them to do a double upgrade.

I thought about including the updated `system.conf` into the new bundles, and adding some code in the bundle hooks that replaces the target's `system.conf` with the new one. However, I found that the bundle hooks don't even get to run because the error I mentioned above happens before bundle hooks are run, and it seems like system-side handlers (which would contain old code) run before bundle-side hooks.

Would anyone who has more knowledge than me on hooks be able to provide insight on how to handle this new slot problem more elegantly?

**Comment by ejoerns** [comment]
@erickson-ureta as you mentioned, the standard way of adding a previously 'forgotten' slot to a system is to upgrade the rootfs first.

The reason is, if you have not described the slot(s) in the `system.conf`, there is no way for RAUC to figure out if the slot is active or inactive and thus if the bundle is installable (before actually starting the installation). A replacement of the `system.conf` currently always requires to restart the rauc service since the config is parsed only once.

Another option for you could be to just update the slot manually by adding a pre- or post-install bundle hook to your rootfs image manifest entry.
Since you can simply pack additional artifacts into a bundle, you could add an image for your new 'slot' and perform the required installation steps from within the hook. Maybe this should be combined with shipping a new `system.conf` to be prepared for future upgrade.

Would this fit your requirements?
